### DISCUSSION TITLE
reload doesn't work from programmatic launch (asyncio)

### DISCUSSION THREAD

**Discussion by SamuelMarks** [Kludex/uvicorn#2144] [Q&A]
### Initial Checks

- [X] I confirm this was discussed, and the maintainers suggest I open an issue.
- [X] I'm aware that if I created this issue without a discussion, it may be closed without a response.

### Discussion Link

```Text
N/A
```

### Description

reload doesn't work

### Example Code

Given something simple like:
```Python
async def serve(reload=True):
    reload_dirs = [path.dirname(__file__)] if reload else None
    config = uvicorn.Config(
        "main:app",
        log_level="info",
        port=4000,
        reload=reload,
        reload_dirs=reload_dirs,
        reload_includes=reload_dirs,
    )
    server = uvicorn.Server(config)
    await server.serve()


if __name__ == "__main__":
    asyncio.run(serve())
```

Output looks good, but it never reloads, e.g., when I change a route's return value:

```
INFO:     Will watch for changes in these directories: ['/tmp/myapp']
INFO:     Started server process [20312]
INFO:     Waiting for application startup.
INFO:     Application startup complete.
INFO:     Uvicorn running on http://localhost:4000 (Press CTRL+C to quit)
```

### Python, Uvicorn & OS Version

```Text
Running uvicorn 0.23.2 with CPython 3.12.0 on Linux
```

> [!IMPORTANT]
> - We're using [Polar.sh](https://polar.sh/encode) so you can upvote and help fund this issue.
> - We receive the funding once the issue is completed & confirmed by you.
> - Thank you in advance for helping prioritize & fund our backlog.

<a href="https://polar.sh/encode/uvicorn/issues/2143">
<picture>
  <source media="(prefers-color-scheme: dark)" srcset="https://polar.sh/api/github/encode/uvicorn/issues/2143/pledge.svg?darkmode=1">
  <img alt="Fund with Polar" src="https://polar.sh/api/github/encode/uvicorn/issues/2143/pledge.svg">
</picture>
</a>

**Comment by james-emerton** [comment]
I am also seeing this issue. A quick look in the code suggests that it is not supported when calling `Server.run()` directly. See the implementation of `uvicorn.run()` here: https://github.com/encode/uvicorn/blob/5bf788f0eb0fc771f5c4eed8f282c7ec256565d2/uvicorn/main.py#L570-L572

I don't think this is currently possible; when `reload=True` the parent process creates the socket and passes it to a child process that terminates and restarts whenever there is a change. If you're starting with a running event loop, it's going to be in the parent process rather than the process that actually ends up serving requests.

In theory, you *could* accomplish this when starting with `Config` and `Server` instances by duplicating what's in `uvicorn.run()`. Indeed you could do the same from an async context with a bit of work, but it isn't likely to accomplish what you're looking for.

I think a small tweak to the [documentation](https://www.uvicorn.org/#config-and-server-instances) would be helpful. It could also be argued that `reload` should not be a `Config` parameter, as it will only be used by `uvicorn.run()`
